src/workerd/api/urlpattern.c++

```diff
@@ -4,309 +4,252 @@
 
 #include "urlpattern.h"
 
-#include <kj/vector.h>
+#include "ada.h"
 
 namespace workerd::api {
+std::optional<URLPattern::URLPatternRegexEngine::regex_type> URLPattern::URLPatternRegexEngine::
+    create_instance(std::string_view pattern, bool ignore_case) {
+  jsg::Lock& js = jsg::Lock::from(v8::Isolate::GetCurrent());
+  jsg::Lock::RegExpFlags flags = jsg::Lock::RegExpFlags::kUNICODE_SETS;
+  if (ignore_case) {
+    flags = static_cast<jsg::Lock::RegExpFlags>(
+        flags | static_cast<int>(jsg::Lock::RegExpFlags::kIGNORE_CASE));
+  }
+
+  return js.tryCatch([&]() -> std::optional<regex_type> {
+    return jsg::JsRef(js, js.regexp(kj::StringPtr(pattern.data(), pattern.size()), flags));
+  }, [&](auto reason) -> std::optional<regex_type> { return std::nullopt; });
+}
+
+bool URLPattern::URLPatternRegexEngine::regex_match(
+    std::string_view input, const regex_type& pattern) {
+  jsg::Lock& js = jsg::Lock::from(v8::Isolate::GetCurrent());
+  return pattern.getHandle(js).match(js, kj::StringPtr(input.data(), input.size()));
+}
 
-namespace {
-jsg::JsRef<jsg::JsRegExp> compileRegex(
-    jsg::Lock& js, const jsg::UrlPattern::Component& component, bool ignoreCase) {
-  return js.tryCatch([&] {
-    jsg::Lock::RegExpFlags flags = jsg::Lock::RegExpFlags::kUNICODE;
-    if (ignoreCase) {
-      flags = static_cast<jsg::Lock::RegExpFlags>(
-          flags | static_cast<int>(jsg::Lock::RegExpFlags::kIGNORE_CASE));
+std::optional<std::vector<std::optional<std::string>>> URLPattern::URLPatternRegexEngine::
+    regex_search(std::string_view input, const regex_type& pattern) {
+  jsg::Lock& js = jsg::Lock::from(v8::Isolate::GetCurrent());
+  KJ_IF_SOME(matches, pattern.getHandle(js)(js, kj::StringPtr(input.data(), input.size()))) {
+    std::vector<std::optional<std::string>> results;
+    results.reserve(matches.size());
```

Comment:
I think we are reserving one more slot than we need here?

Reply:
Yes but adding -1 adds a branch (checking if size is bigger than 0), which reduces the impact of this optimization